presto-main/src/main/java/com/facebook/presto/execution/scheduler/LegacySqlQueryScheduler.java

```diff
@@ -459,9 +505,182 @@ else if (!result.getBlocked().isDone()) {
                 // get all sections ready for execution
                 .filter(this::isReadyForExecution)
                 .limit(maxConcurrentMaterializations - runningPlanSections)
+                .map(this::tryCostBasedOptimize)
                 .collect(toImmutableList());
     }
 
+    /**
+     * Utility function to invoke runtime cost-based optimizer, which determines if the probe and build side of a JoinNode should be swapped,
+     * based on the statistics of the temporary table holding materialized exchange outputs from finished children sections.
+     */
+    private StreamingPlanSection tryCostBasedOptimize(StreamingPlanSection section)
+    {
+        // no need to do runtime optimization if no materialized exchange data is utilized by the section.
+        if (!isRuntimeOptimizerEnabled(session) || getExchangeMaterializationStrategy(session) != ALL || section.getChildren().isEmpty()) {
+            return section;
+        }
+
+        // Apply runtime optimization on each StreamingSubPlan's fragment
+        Map<PlanFragment, PlanFragment> oldToNewFragment = new HashMap<>();
+        stream(forTree(StreamingSubPlan::getChildren).depthFirstPreOrder(section.getPlan()))
+                .map(StreamingSubPlan::getFragment)
+                .forEach(fragment -> {
+                    PlanNode newRoot = fragment.getRoot();
+                    for (PlanOptimizer optimizer : runtimePlanOptimizers) {
+                        newRoot = optimizer.optimize(newRoot, session, variableAllocator.getTypes(), variableAllocator, idAllocator, warningCollector);
+                    }
+                    // The partitioningScheme should stay the same
+                    // even if the root's outputVariables are changed.
```

Comment:
I wonder if the runtime plan optimizers should operate on the whole fragment rather than just the plan. That would be so that in the future we can add optimizers that change the partitioning.  We can save that though for when we have any such optimizers.